src/CommandLine.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "CommandLine.h"
+
+static void
+ResetGetopt(const bool allowStderrWarnings)
+{
+    opterr = allowStderrWarnings ? 1 : 0;
+    // getopt(3) uses global state but resets it if optind is zero
+    // getopt(3) always skips argv[0], even if optind is zero
+    optind = 0;
+}
+
+CommandLine::CommandLine(int argC, char *argV[], const char *shortRules, const struct option *longRules):
+    argv_(),
+    shortOptions_(xstrdup(shortRules)),
+    longOptions_()
+{
+    assert(argC > 0); // C++ main() requirement that makes our arg0() safe
+
+    /* copy argV items */
+    argv_.reserve(argC+1);
+    for (int i = 0; i < argC; ++i)
+        argv_.push_back(xstrdup(argV[i]));
+    argv_.push_back(nullptr); // POSIX argv "must be terminated by a null pointer"
+
+    /* copy grammar rules for the long options */
+    for (auto longOption = longRules; longOption; ++longOption) {
+        longOptions_.push_back(*longOption);
+        if (!longOption->name)
+            break;
+    }
+}
+
+CommandLine::CommandLine(const CommandLine &them):
+    CommandLine(them.argc(), them.argv(), them.shortOptions_, them.longOptions())
+{
+}
+
+CommandLine &
+CommandLine::operator =(CommandLine them) // not a reference so that we can swap
```

Comment:
Is it better to use more natural `CommandLine::operator =(const CommandLine &them)` and add one more instruction to copy item into a temporary local variable?

Reply:
Done.